README.md
# discord-haskell-monad

This is a supplementary library to `discord-haskell` that provides an
exception-oriented abstraction layer for Discord API REST calls. It bases its
existence on the idea that Either should be reserved for pure code, while
Exceptions should be used for IO due to its complex unpredictability. This idea
comes from various sources including the
[Exceptions Best Practices](https://www.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/)
post by Michael Snoyman and the
[Learn Haskell by Building a Blog Generator](https://lhbg-book.link/06-errors_and_files/05-summary.html)
book by Gil Mizrahi.

Reasoning about errors in IO with Either, as `discord-haskell` currently does
with `RestCallErrorCode`, is unreliable and leads to unnecessarily messy user
code. The line grows blurry between what is handled by the Either and what is not.
For instance, while e.g. HTTP 400 errors are returned by `restCall` through
`Left RestCallErrorCode`, connection timeout exceptions are not. Many are logged
to `discord-haskell`'s internal log. I wondered whether those HTTP 400 errors
truly needed to be handled with Either anyway.

The library, attempting to be an answer to that question, introduces a
`MonadDiscord` type-class. It has a single method: `call`. This is a sort-of
replacement of `restCall`, but throws HTTP error codes, network errors, and
parse errors as Exceptions in IO.

```hs
example :: (MonadDiscord m) => m ()
example = do
    m <- call (R.CreateMessage 123123123 "Hello World!")
-- `catches` [Handler (\ (ex :: RestCallErrorCode)      -> handleAPIError ex),
--            Handler (\ (ex :: ResponseParseException) -> handleParseError ex),
--            Handler (\ (ex :: HttpException)          -> handleHttpError ex)]
    print $ messageId m
```

The library also has two instances of this monad defined, one for
`DiscordHandler` and one for IO environments with access an auth token.

## The DiscordHandler instance

The DiscordHandler instance of `DiscordMonad` solves the indentation problem
that is present with the `Either RestCallErrorCode a` type that `discord-haskell`
uses for its calls. For example, the following is what you might typically write
with `discord-haskell`.

```hs
-- Send a message and react on it
example :: DiscordHandler ()
example = do
  eMsg <- restCall $ R.CreateMessage 123412341234 "Hello World!"
  case eMsg of
    Left e -> liftIO (print "first msg failed") >> pure ()
    Right msg -> do
      result <- restCall $ R.CreateReaction (messageChannel msg, messageId msg) "eyes"
      case result of
        Left e -> liftIO (print "reaction failed") >> pure ()
        Right _ -> liftIO $ putStrLn "All good!"
```

You could use the EitherT transformer, but if rollback is required for certain
actions it is difficult. The code is also more difficult to understand since you
have to sprinkle `lift`s throughout.

The following is what you would replace the above by, using this library.

```hs
import Discord.Monad
--- Send a message and react on it
example :: DiscordHandler ()
example = do
  msg <- call $ R.CreateMessage 123412341234 "Hello World!"
  call $ R.CreateReaction (messageChannel msg, messageId msg) "eyes"
  liftIO $ putStrLn "All good!"
```

The `RestCallErrorCode` that was previously in `Left` is now an Exception.
This means that you can use `handle`, `catch`, `finally`, etc (from the
`safe-exceptions` package) wherever.

In addition, the code may now explicitly throw a `HttpException` or a
`ResponseParseException`, which were previously consumed after logging. Note that,
due to this separation of errors and explicit exception throws, the functionality
you get from `try . call $ ... :: Either RestCallErrorCode a` is different to
the original `restCall`. Specifically, all parse errors which would have
originally been returned as ErrorCode 400 will not be caught by that `try`, nor
will HttpExceptions, which would originally have caused an automatic retry of
the request.

## The MonadReader Auth instance

This library also provides an instance of `MonadDiscord` for all Monads `m` that
satisfy the constraint `(MonadIO m, MonadReader Auth m, MonadMask m)`.

In plain English, this means any IO monad with access to the Auth can send
Discord requests (complete with rate-limiting). This does not require running
the Bot's Rest and Event loop, reducing the resources needed to send a simple
API request.

```hs
main :: IO Message
main = runReaderT (call $ R.CreateMessage 123412341234 "Hello World!") (Auth "token here")
```

## Concluding Notes

Please feel free to contribute through issues or pull requests if you want to add anything!

This library was developed while developing a [Discord bot for our university server](https://github.com/yellowtides/owenbot-hs).


Changelog.md
# Changelog

## 1.2.1

- Allow `discord-haskell` up to 1.14.0 using CPP conditionals

## 1.2.0

- Breaking: Remove all functions for Rest commands as part of re-assessing the raison d'être of the library. The minimal and only definition is `call`, replacing `processRequest`.
- Throw `HttpException` properly on `HttpException`s (and not `RestCallErrorCode`).
- Throw `ResponseParseException` on parse errors instead of `RestCallErrorCode`.
- Relax constraints on the Reader instance, to `MonadIO m, MonadReader Auth m => MonadDiscord m`.
- Allow `discord-haskell` up to 1.13.0

## 1.1.0

- Bump Stack LTS to 18.24
- Bump `discord-haskell` to 1.12.0
- Add functions in `DiscordMonad` for Interactions and some changes with GuildWidget
- Introduce `processRequest` as a minimal definition for the `DiscordMonad` class, making ReaderT and DiscordHandler instances simpler.

## 1.0.0

- Initial release.
- `DiscordMonad` typeclass with all possible Rest commands defined as functions.
- `MonadIO m => ReaderT Auth m` and `DiscordHandler` instances for `DiscordMonad`.


src/Discord/Internal/Monad.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE FlexibleInstances #-} -- allow instance declaration of MonadDiscord
{-# LANGUAGE FlexibleContexts #-} -- allow instance declaration for MonadReader Auth
{-# LANGUAGE UndecidableInstances #-} -- necessary for MonadIO constraint on MonadDiscord
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE CPP #-}
{-|
Module      : Discord.Internal.Monad
Description : The DiscordMonad class and its instances.
Copyright   : (c) 2022 Yuto Takano
Maintainer  : moa17stock@gmail.com
License     : MIT (see the LICENSE file)

This module contains the 'MonadDiscord' type class to call Discord Requests.
An instance for 'DiscordHandler' and all monads that satisfy @MonadReader Auth@
are provided.

Based on the idea that Either should be used only for pure code and that
Exceptions should be used for unpredictable IO code, this library puts a thin
layer on top of @discord-haskell@ so that all types of errors are thrown as
exceptions. This includes: 'RestCallErrorCode', 'Req.HttpException', and
'ResponseParseException'.

A bit of an example code of what this means:

@@
example :: (MonadDiscord m) => m ()
example = do
    m <- call (R.CreateMessage 123123123 "Hello World!")
-- `catches` [Handler (\ (ex :: RestCallErrorCode)      -> handleAPIError ex),
--            Handler (\ (ex :: ResponseParseException) -> handleParseError ex),
--            Handler (\ (ex :: HttpException)          -> handleHttpError ex)]
    print $ messageId m
@@

Please see the documentation below, as well as the README on GitHub for
explanations.
-}

module Discord.Internal.Monad where

import Control.Concurrent (threadDelay, isEmptyMVar)
import Control.Exception.Safe (Exception, MonadMask, throwIO)
import Control.Monad (void)
import Control.Monad.IO.Class (MonadIO, liftIO)
import Control.Monad.Reader (MonadReader, ask)
import qualified Data.Aeson as A (FromJSON, eitherDecode)
import qualified Data.ByteString.Lazy as BL
import qualified Data.Text.Encoding as TE
import qualified Data.Text as T
import qualified Network.HTTP.Req as Req

import Discord.Internal.Rest (writeRestCall)
import Discord.Internal.Rest.Prelude
import Discord.Internal.Rest.HTTP
    (RestCallInternalException(..), Request(..), JsonRequest(..))
import Discord.Handle (DiscordHandle(..))
import Discord.Types
import Discord

#if MIN_VERSION_discord_haskell(0,14,0)
#else
-- | We redefine the 'RestCallErrorCode' as an IO Exception that is thrown when
-- the Discord API returns an unexpected response code.
instance Exception RestCallErrorCode
#endif

-- | We introduce a new datatype for parse errors. This replaces discord-haskell's
-- default behaviour which is to print the error to the log and return a 400
-- rest error code.
data ResponseParseException = ResponseParseException String BL.ByteString
  deriving Show
instance Exception ResponseParseException

-- | @MonadDiscord@ is a class of Monads that can invoke a Discord
-- 'Discord.Internal.Rest.Prelude.Request'.
class (Monad m, MonadMask m) => MonadDiscord m where
    {-# MINIMAL call #-}
    -- | Calls the Discord API with the specific Request.
    -- This may throw the following:
    --
    -- * 'RestCallErrorCode' on some Discord error
    -- * 'Req.HttpException' on HTTP error
    -- * 'ResponseParseException' on a parse error
    --
    -- In addition, any instances of MonadDiscord should obey and throw only
    -- exceptions of the above type in appropriate cases.
    call :: (Request (r a), A.FromJSON a) => r a -> m a

-- | 'DiscordHandler' is trivially a Discord monad.
-- We recreate the implementation of 'restCall' but slightly altered to rethrow
-- exceptions in the style we want.
--
-- See the implementation of restCall here:
-- https://github.com/aquarial/discord-haskell/blob/c009edc5a5cd4991f600315686d5a1c880ea683a/src/Discord.hs#L128
instance {-# OVERLAPPING #-} MonadDiscord DiscordHandler where
    call req = do
        h <- ask
        empty <- liftIO $ isEmptyMVar (discordHandleLibraryError h)
        if not empty then
            throwIO $ RestCallErrorCode 400 "Library Stopped Working" ""
        else do
            -- writeRestCall and discord-haskell's internals handle rate limits
            -- and retry automatically.
            resp <- liftIO $ writeRestCall (discordHandleRestChan h) req
            case resp of
                Right x -> pure x
                Left (RestCallInternalErrorCode c status body) ->
                    throwIO $ RestCallErrorCode c (TE.decodeUtf8 status) (TE.decodeUtf8 body)
                Left (RestCallInternalHttpException e1) ->
                    -- in discord-haskell's restCall, the request is retried here
                    -- automatically after a delay. We choose not to do that and
                    -- rethrow the HttpException.
                    throwIO e1
                Left (RestCallInternalNoParse err dat) ->
                    throwIO $ ResponseParseException err dat

-- | Any Reader monad with the 'Auth' accessible is a valid Discord monad.
--
-- Usage:
--
-- @
-- import Control.Monad.Reader (runReaderT)
-- someFunc :: IO Message
-- someFunc = runReaderT (call $ CreateMessage 1234 "text") (Auth "tokenhere")
-- @
instance (Monad m, MonadIO m, MonadMask m, MonadReader Auth m) => MonadDiscord m where
    -- The following is an adapted and simplified version of the logic within
    -- discord-haskell's @Discord.Internal.Rest.HTTP@ module. It does not handle
    -- HTTPExceptions, and also throws all RestCallErrorCodes immediately upon
    -- receiving. The function will retry if a rate limit is encountered.
    call req = do
        -- get the token
        auth <- ask
        -- create a request with the auth in the header
        let action = compileRequest auth (jsonRequest req)
        -- send off the request, get the response. May throw HttpException.
        resp <- liftIO $ restIOtoIO action
        let body = Req.responseBody resp
            code = Req.responseStatusCode resp
            status = TE.decodeUtf8 $ Req.responseStatusMessage resp
        case () of
            _ | code `elem` [429, 500, 502] -> do
                -- wait a bit before retrying.
                liftIO $ threadDelay (10 * 10 ^ (6 :: Int))
                call req
            _ | code >= 200 && code <= 299 -> do
                let parsableBody = if body == "" then "[]" else body
                case A.eitherDecode parsableBody of
                    Right o -> pure o
                    Left er -> throwIO $ ResponseParseException er body
            _ | otherwise -> do
                throwIO $ RestCallErrorCode
                    code
                    status
                    (TE.decodeUtf8 $ BL.toStrict body)

-- | From a 'JsonRequest', create a RestIO thingy with the appropriately auth
-- headers.
compileRequest :: Auth -> JsonRequest -> RestIO Req.LbsResponse
compileRequest auth request =
    let authopt = authHeader auth <> Req.header "X-RateLimit-Precision" "millisecond"
    in
        case request of
            (Delete url opts) ->
                Req.req Req.DELETE url Req.NoReqBody Req.lbsResponse (authopt <> opts)
            (Get url opts) ->
                Req.req Req.GET url Req.NoReqBody Req.lbsResponse (authopt <> opts)
            (Put url body opts) ->
                Req.req Req.PUT url body Req.lbsResponse (authopt <> opts)
            (Patch url body opts) -> do
                b <- body
                Req.req Req.PATCH url b Req.lbsResponse (authopt <> opts)
            (Post url body opts) -> do
                b <- body
                Req.req Req.POST url b Req.lbsResponse (authopt <> opts)

